tensorflow/core/util/gpu_device_functions.h

```diff
@@ -304,61 +332,93 @@ __device__ inline T CudaShuffleDownSync(unsigned mask, T value, unsigned delta,
 // Variant of the (undocumented) version from the CUDA SDK, but using unsigned
 // instead of float for lo and hi (which is incorrect with ftz, for example).
 // See b/69446944.
-__device__ inline double CudaShuffleDownSync(unsigned mask, double value,
+__device__ inline double GpuShuffleDownSync(unsigned mask, double value,
                                              unsigned delta,
                                              int width = warpSize) {
   unsigned lo, hi;
+#if GOOGLE_CUDA
   asm volatile("mov.b64 {%0,%1}, %2;" : "=r"(lo), "=r"(hi) : "d"(value));
-  hi = CudaShuffleDownSync(mask, hi, delta, width);
-  lo = CudaShuffleDownSync(mask, lo, delta, width);
+  hi = GpuShuffleDownSync(mask, hi, delta, width);
+  lo = GpuShuffleDownSync(mask, lo, delta, width);
   asm volatile("mov.b64 %0, {%1,%2};" : "=d"(value) : "r"(lo), "r"(hi));
   return value;
+#elif TENSORFLOW_USE_ROCM
+  uint64_t tmp = static_cast<uint64_t>(value);
+  lo = static_cast<unsigned>(tmp);
+  hi = static_cast<unsigned>(tmp >> 32);
+  hi = __shfl_down(static_cast<int>(hi), delta, width);
+  lo = __shfl_down(static_cast<int>(lo), delta, width);
+  return static_cast<double>(static_cast<uint64_t>(hi) << 32 | static_cast<uint64_t>(lo));
+#endif
 }
 
 // Wrapper for __shfl_xor_sync. All threads in 'mask' must call this function in
 // convergence, see comment above for details.
 template <typename T>
-__device__ T CudaShuffleXorSync(unsigned mask, T value, int lane_mask,
+__device__ T GpuShuffleXorSync(unsigned mask, T value, int lane_mask,
```

Comment:
@chsigg Please refer here to see ROCm-specific logic for `GpuShuffleXorSync` for various data types.